vmm/src/memory_manager.rs

```diff
@@ -453,14 +464,33 @@ impl MemoryManager {
         Ok(slot)
     }
 
+    pub fn balloon_resize(&mut self, size: u64) -> Result<(), Error> {
+        if let Some(balloon) = &self.balloon {
+            balloon
+                .lock()
+                .unwrap()
+                .resize(size)
+                .map_err(Error::VirtioBalloonResizeFail)?;
+            self.balloon_size = size;
+        }
+
+        Ok(())
+    }
+
     pub fn resize(&mut self, desired_ram: u64) -> Result<bool, Error> {
-        if desired_ram > self.current_ram {
-            self.hotplug_ram_region((desired_ram - self.current_ram) as usize)?;
-            self.current_ram = desired_ram;
-            Ok(true)
-        } else {
-            Ok(false)
+        if desired_ram >= self.current_ram {
+            if self.balloon_size != 0 {
+                self.balloon_resize(0)?;
+            }
+            if desired_ram > self.current_ram {
+                self.hotplug_ram_region((desired_ram - self.current_ram) as usize)?;
+                self.current_ram = desired_ram;
+                return Ok(true);
+            }
+        } else if desired_ram < self.current_ram {
+            self.balloon_resize(self.current_ram - desired_ram)?;
         }
+        Ok(false)
```

Comment:
I'm really confused about the way balloon is supposed to be handled, and that means I'm not sure if we should use the balloon along with the memory hotplug by default.
Usually the way it works, the management layer such as Kata Containers will ask for the balloon to be extended, this shouldn't be up to the VMM to choose this kind of thing.
So I think we might need a dedicated command to resize the balloon only.
/cc @rbradford @sameo

Reply:
I agree we should not use ballooning for the `resize` operation on the VM as it's effectively not resizing the VM but releasing some pages to the host. The VM memory size is not changed.
As you suggested, ballooning requests should come from higher levels of the stack, through a dedicated API that we should provide.

Reply:
Could you please give me some advises about the control interface of the balloon?

Reply:
We'd need a `/api/v1/vm.balloon` command that would take a size argument to set the new balloon size:

```Patch
diff --git a/vmm/src/api/openapi/cloud-hypervisor.yaml b/vmm/src/api/openapi/cloud-hypervisor.yaml
index c16e05a5..80d25af3 100644
--- a/vmm/src/api/openapi/cloud-hypervisor.yaml
+++ b/vmm/src/api/openapi/cloud-hypervisor.yaml
@@ -219,6 +219,22 @@ paths:
         500:
           description: The new device could not be added to the VM instance.
 
+  /vm.balloon:
+    put:
+      summary: Resize the VM virtio balloon device
+      requestBody:
+        description: The target size for the balloon device
+        content:
+          application/json:
+            schema:
+              $ref: '#/components/schemas/VmBalloonResize'
+        required: true
+      responses:
+        204:
+          description: The VM balloon device was successfully resized.
+        404:
+          description: The VM balloon device could not be resized.
+
 components:
   schemas:
 
@@ -542,3 +558,11 @@ components:
       properties:
         id:
           type: string
+
+    VmBalloonResize:
+      type: object
+      properties:
+        size:
+          description: requested ballon size in bytes
+          type: integer
+          format: int64
```

Reply:
Added.